### ISSUE TITLE
Google Cloud Tasks with Google App Engine: 14 UNAVAILABLE: read ECONNRESET

### ISSUE THREAD

**Issue by AllanOliveiraM** [googleapis/google-cloud-node#5852]
### Discussed in https://github.com/googleapis/google-cloud-node/discussions/5799

<div type='discussions-op-text'>

<sup>Originally posted by **AllanOliveiraM** November 12, 2024</sup>
Hi everyone, how's it going?

I need some help with the Cloud Tasks client library. I have an application running on App Engine, and I use Cloud Tasks to queue the processing of some critical application data.

I'm using the `createTask` method to do this, and everything works normally most of the time. The problem is that occasionally, a task cannot be created, and I encounter a "14 UNAVAILABLE: read ECONNRESET" error in the logs.

I believe the issue is not with the API itself, but with my client library configuration.

I found [this issue](https://github.com/grpc/grpc-node/issues/1907) in the grpc-node GitHub repository, but I'm still having trouble configuring the client correctly.

If anyone could help me understand what I’m doing wrong, I’d be very grateful :)

![image](https://github.com/user-attachments/assets/fe6f9248-7d76-48d2-a8ee-2d1840469d96)

I don’t think the error is here, but I’ll include it as context.

![image](https://github.com/user-attachments/assets/a0bd88bf-0f6d-414e-8a07-2048071db1b6)
![image](https://github.com/user-attachments/assets/70cb7c41-b99c-429f-ba6b-09711f70b883)

</div>

**Comment by sofisl** [comment]
- What version are you using of:
    - Node
    - tasks
    - gax
    - grpc-js?
- Does this issue happen outside of using NestJs (can you recreate it just with AppEngine)?
- Is there any opportunity to reuse an existing CloudTasksClient in your code instead of creating a new one (perhaps its entirely optimized, I don't know, but I wonder if it has anything to do with creating new clients).
- Since you're using NestJS, are you sure you're not using `grpc` (the deprecated package)?
- Did you enable [logging](https://github.com/grpc/grpc-node/blob/34b82cbf7ffacbb717f3796eac686085ce16dc3d/TROUBLESHOOTING.md?plain=1#L8)?

**Comment by BenjaminDish** [comment]
I'm facing the exact same trouble than @AllanOliveiraM on my app.

I'm using this library to run Google cloud jobs. The calls are made every 5 minutes, triggered by a cron table, and >99% of them works well. But sometimes, randomly, I get this error.

The method I'm calling is :
```
import { JobsClient, protos } from '@google-cloud/run';
const runClient = new JobsClient();

// Construct request
const request: Partial<protos.google.cloud.run.v2.RunJobRequest> = {
  name: myJobName,
};

// Run request
await this.runClient.runJob(request);
```

And sometimes, (< 1% of the calls) I get this error : 

`[ExceptionsHandler] error: 14 UNAVAILABLE: read ECONNRESET`

No more logs in my back-end. There are no footprints of this call on GCP side.

**Comment by AllanOliveiraM** [comment]
I'm compiling better data for debugging and feedback shortly.

**Comment by AllanOliveiraM** [comment]
# More info

Node: v23.7.0

## Used package versions:

- `@google-cloud/tasks`: 5.5.1  
- `google-gax`: 4.4.1  
- `@grpc/grpc-js`: 1.9.9 (this version is internally resolved by Yarn through `google-gax`)

I'm already reusing the same client instance across the project. In NestJS, I'm using a singleton class where the client is instantiated only once.

I'm also sure that I'm not using the deprecated `grpc` package because I'm directly using the `@google-cloud/tasks` library, which depends on a recent version of `google-gax`.

I also use other Google libraries in the project, but I believe none of them should interfere with this:

```json
{
    "@google-cloud/logging-winston": "^6.0.0",
    "@google-cloud/secret-manager": "^5.6.0",
    "@google-cloud/storage": "^7.14.0",
    "@google-cloud/tasks": "^5.5.1",
    "@google/generative-ai": "^0.21.0"
}
```

---

Other libraries installed directly in the project:

- `@grpc/grpc-js`: ^1.10.10  
- `@grpc/proto-loader`: ^0.7.13  
- `@nestjs/axios`: ^3.0.1  
- `@nestjs/cache-manager`: ^2.1.0  
- `@nestjs/common`: ^10.2.3  
- `@nestjs/config`: ^3.0.1  
- `@nestjs/core`: ^10.2.3  
- `@nestjs/microservices`: ^10.3.10  
- `@nestjs/platform-express`: ^10.4.8  
- `@nestjs/terminus`: ^10.0.1  
- `@nestjs/throttler`: ^4.2.1  

I also use gRPC directly in another service — this is a monorepo.  
There are 4 apps in total, and the one experiencing the issue is **the one that does not use `@grpc/grpc-js` directly**. The affected app only uses `@google-cloud/tasks`.

I have already enabled logging, but the only log that appears is the error mentioned above.

---

After spending a lot of time trying to understand what’s happening, it seems to me that there might be some kind of connection timeout — specifically when running on AppEngine. However, I couldn’t find a way to configure reconnection options.

I came across this in another issue (the one I mentioned above). [Please, take a look there](https://github.com/grpc/grpc-node/issues/1907#issuecomment-1399670912). I found this configuration for gRPC, but I couldn’t figure out how to apply these settings when using `@google-cloud/tasks`:

```typescript
const channelOptions: ChannelOptions = {
  // Send keepalive pings every 10 seconds, default is 2 hours.
  'grpc.keepalive_time_ms': 10 * 1000,
  // Keepalive ping timeout after 5 seconds, default is 20 seconds.
  'grpc.keepalive_timeout_ms': 5 * 1000,
  // Allow keepalive pings when there are no gRPC calls.
  'grpc.keepalive_permit_without_calls': 1,
};
```



I wasn’t able to reproduce this issue outside the production environment at all.

**Comment by AllanOliveiraM** [comment]
@sofisl I can't reopen the issue 😑

**Comment by AllanOliveiraM** [comment]
full code in my project:

```ts
import { HttpService } from '@nestjs/axios'
import {
  HttpStatus,
  Injectable,
  InternalServerErrorException,
  Logger,
} from '@nestjs/common'
import { ConfigService } from '@nestjs/config'

import { CloudTasksClient } from '@google-cloud/tasks'
import { captureException } from '@sentry/node'
import { AxiosError } from 'axios'
import { catchError, firstValueFrom } from 'rxjs'

import { GlobalEnvVariables } from '@app/common/constants/environment'
import { AvailableQueueDefinition } from '@app/common/definitions/queues'

@Injectable()
export class TaskManagerService {
  private readonly logger = new Logger(TaskManagerService.name)
  private cloudTasksClient: CloudTasksClient | null = null
  private googleCloudProjectId: string | null
  private googleCloudTasksLocation: string | null
  private isCloudTasksEnabled: boolean

  constructor(
    private readonly configService: ConfigService,
    private readonly httpService: HttpService
  ) {
    this.googleCloudProjectId =
      this.configService.get(GlobalEnvVariables.GOOGLE_CLOUD_PROJECT_ID) || null
    this.googleCloudTasksLocation =
      this.configService.get(GlobalEnvVariables.GOOGLE_CLOUD_TASKS_LOCATION) || null

    this.isCloudTasksEnabled = (() => {
      if (this.configService.getOrThrow(GlobalEnvVariables.NODE_ENV) !== 'production') {
        return false
      }

      if (!this.configService.getOrThrow(GlobalEnvVariables.GOOGLE_CLOUD_TASKS_ENABLED)) {
        return false
      }

      return true
    })()

    if (this.isCloudTasksEnabled) {
      this.cloudTasksClient = new CloudTasksClient()
    }
  }

  private async createGoogleCloudTask({
    queue,
    payload,
  }: {
    queue: AvailableQueueDefinition
    payload: Record<string, any>
  }) {
    if (
      [
        this.cloudTasksClient,
        this.googleCloudProjectId,
        this.googleCloudTasksLocation,
      ].some(val => !val)
    ) {
      const err = new InternalServerErrorException({
        statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
        message: 'Google APIs not enabled. ERR: GCT',
      })

      captureException(err, {
        extra: {
          envs: [
            this.cloudTasksClient,
            this.googleCloudProjectId,
            this.googleCloudTasksLocation,
          ],
          queue,
        },
      })

      throw err
    }

    const parent = this.cloudTasksClient.queuePath(
      this.googleCloudProjectId,
      this.googleCloudTasksLocation,
      queue.queueName
    )

    const task = {
      appEngineHttpRequest: {
        headers: {
          'Content-Type': 'application/json',
        },
        httpMethod: 'POST' as const,
        relativeUri: queue.uri,
        body: Buffer.from(JSON.stringify(payload)).toString('base64'),
      },
    }

    try {
      await this.cloudTasksClient.createTask({
        parent,
        task,
      })
    } catch (error) {
      this.logger.error(error)

      captureException(error, {
        extra: {
          queue,
          payload,
        },
      })

      throw new InternalServerErrorException({
        statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
        message: 'Task Manager failed to receive a task.',
      })
    }
  }

  async createTask({
    queue,
    payload,
  }: {
    queue: AvailableQueueDefinition
    payload: Record<string, any>
  }) {
    if (this.isCloudTasksEnabled) {
      await this.createGoogleCloudTask({
        queue,
        payload,
      })

      return
    }

    if (this.configService.getOrThrow(GlobalEnvVariables.NODE_ENV) === 'development') {
      await firstValueFrom(
        this.httpService
          .post(queue.uri, payload, {
            headers: {
              'x-appengine-queuename': queue.queueName,
            },
          })
          .pipe(
            catchError((error: AxiosError) => {
              this.logger.error('[DEV TASK MANAGER ERROR] ', error.response.data)

              throw new InternalServerErrorException({
                statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
                message: 'Development task manager error.',
                _responseData: error.response.data,
              })
            })
          )
      )

      return
    }

    captureException(
      new InternalServerErrorException({
        statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
        message: 'No Task Manager defined to receive tasks.',
      }),
      {
        extra: {
          queue,
          payload,
        },
      }
    )

    throw new InternalServerErrorException({
      statusCode: HttpStatus.INTERNAL_SERVER_ERROR,
      message: 'No Task Manager defined to receive tasks.',
    })
  }
}

```

**Comment by omar-dev-amrk** [comment]
I'm facing this issue too, any workarounds and best practices? is it safe to implement retry for this exact error code?

**Comment by AllanOliveiraM** [comment]
@omar-dev-amrk 

I'm testing a hypothesis:

I'm running my server on Google App Engine, and I recently remembered reading that App Engine has a one-hour limit for active connections, as far as I recall.

The issue is that I'm not sure if this limit also applies to *outgoing* connections within GCP itself. From what I understand in the documentation, this limit applies to incoming TCP connections in general — and therefore to HTTP connections.

I'm suspecting that the `@google-cloud/tasks` library, via `google-gax`, via `@grpc/grpc-js`, is configured to keep a persistent connection open.  
Maybe this connection is being kept alive until it hits that one-hour limit — so on the next request, since the existing connection is no longer valid, we get an error.

I'm currently not able to fully work around this issue with confidence. Here's what I did to try to mitigate it:

I had some mechanisms that kept the instance alive almost constantly. I’ve now optimized things so that it only stays active when really necessary. But I suspect this won’t entirely solve the problem, because in practice the instances are still being used almost all the time.

Up to this point, I'm running this in production, and I haven’t seen the error again — but since it occurs unpredictably, I can't be confident that it won't come back.

I’m now implementing a mechanism to recreate the client instance (`CloudTasksClient`) every 30 minutes, hoping this will force the connection to be reopened and prevent further issues. I’m waiting for the error to happen again so I can better understand the problem — if it does happen, I’ll publish this workaround and monitor if the error returns.

I imagine this behavior could be controlled via some gRPC client configuration, but I couldn’t find any way to do that — and no one from the official `grpc-node` repository has responded to this yet. I believe the configuration probably still needs to be done through `google-gax` or `@google-cloud/tasks`, but I haven’t figured out how.

I did find the following configuration for gRPC, which apparently solved the same issue — though it wasn’t used with the Google Tasks library:

```ts
const channelOptions: ChannelOptions = {
  // Send keepalive pings every 10 seconds, default is 2 hours.
  'grpc.keepalive_time_ms': 10 * 1000,
  // Keepalive ping timeout after 5 seconds, default is 20 seconds.
  'grpc.keepalive_timeout_ms': 5 * 1000,
  // Allow keepalive pings when there are no gRPC calls.
  'grpc.keepalive_permit_without_calls': 1,
};
```

If you know how to apply these configurations in this scenario, that would be fantastic.

If you end up implementing any kind of retry mechanism, could you share it here with me? That might give me some insight in case I reach the same point and need to do the same.

In my case, there are some redundancies in place that usually get the job done when the library fails, so I haven’t implemented any retry logic so far.

**Comment by omar-dev-amrk** [comment]
@AllanOliveiraM 
I think you can apply the config above like this:

```js
import { CloudTasksClient } from '@google-cloud/tasks';

const channelOptions = {
  // Send keepalive pings every 10 seconds, default is 2 hours.
  'grpc.keepalive_time_ms': 10 * 1000,
  // Keepalive ping timeout after 5 seconds, default is 20 seconds.
  'grpc.keepalive_timeout_ms': 5 * 1000,
  // Allow keepalive pings when there are no gRPC calls.
  'grpc.keepalive_permit_without_calls': 1,
};

export const cloudTasksClient = new CloudTasksClient({
  channelOptions,
});

```
 I figured that out after digging in the source code. But I haven't tried it yet. Claude research thinks they're problematic. Here's what was found:


# Critical Production Issues in Cloud Tasks Configuration

Your Google Cloud Tasks client configuration contains **several critical production issues** that will cause reliability problems and performance degradation. The most severe issue is the aggressive 10-second keepalive setting, which violates gRPC specifications and will cause intermittent connection failures in production.

**Bottom line**: The current configuration needs immediate changes before production deployment, particularly the keepalive settings which will cause "transport is closing" errors and connection instability.

## Aggressive keepalive settings create connection failures

The **10-second keepalive interval is critically problematic** and violates gRPC's A8-client-side-keepalive specification. This setting will cause production failures through several mechanisms:

**Ping strike enforcement**: Most gRPC servers enforce a minimum keepalive time of 5 minutes (300,000ms). Your 10-second interval triggers "ping strikes" - after 3 aggressive pings, servers terminate connections with GOAWAY messages, causing **UNAVAILABLE/14 errors** with "transport is closing" messages.

**Resource overhead**: The frequent pings consume unnecessary CPU and network resources. Production deployments from Netflix, Datadog, and other large-scale systems use 5-minute intervals specifically to avoid these issues.

**Load balancer conflicts**: Google Cloud Load Balancers maintain connections for 10 minutes by default. Your aggressive client-side keepalives can interfere with load balancer connection management, creating timing conflicts.

**Immediate fix required**: Change `grpc.keepalive_time_ms` from 10,000 to **300,000** (5 minutes) and set `grpc.keepalive_permit_without_calls` to **false** to prevent server coordination issues.



## Production deployment anti-patterns

The configuration demonstrates several production deployment anti-patterns that will impact reliability and performance:

**Missing graceful shutdown**: No cleanup code calls `Close()` methods on client instances. This creates resource leaks and prevents proper connection termination during application shutdown.

**Inadequate error monitoring**: The configuration lacks instrumentation for tracking gRPC-specific errors like ping strikes, connection failures, and retry patterns that are critical for production observability.

**Deployment coordination gaps**: Changes to keepalive settings require careful server-then-client deployment ordering. Your current configuration would fail if deployed without coordinating with backend services.

## Recommended production configuration

Replace your current configuration with this production-ready version:

```javascript
import { CloudTasksClient } from '@google-cloud/tasks';

// Production-safe gRPC settings
const channelOptions = {
  // Conservative 5-minute keepalive (gRPC best practice)
  'grpc.keepalive_time_ms': 300000,
  // 5-second timeout sufficient for ping response
  'grpc.keepalive_timeout_ms': 5000,
  // Disable pings without active calls to avoid server conflicts
  'grpc.keepalive_permit_without_calls': false,
};

// Single reusable client instance
export const cloudTasksClient = new CloudTasksClient({
  channelOptions,
});

// Graceful shutdown handling
process.on('SIGTERM', async () => {
  await cloudTasksClient.close();
  process.exit(0);
});
```

**Key improvements**: Conservative keepalive settings prevent connection failures and added proper cleanup handling.

## Conclusion

The current configuration will cause **production reliability issues** primarily due to aggressive keepalive settings that violate gRPC specifications. The most critical change is increasing the keepalive interval from 10 seconds to 5 minutes. These changes should be implemented before any production deployment to avoid connection failures and resource waste.





Let me know what you think. I'm guessing retries are safer at this point.
